devdoc/waiter.md

```diff
@@ -0,0 +1,71 @@
+# `waiter` requirements
+
+## Overview
+
+`waiter` is a module that allows the user to move data asynchronously from producer to consumer without having to block the thread.
+
+Consumers call `waiter_register_notification` to register a `NOTIFICATION_CALLBACK` function to be called when data becomes available.
+
+Producers call `waiter_notify` to notify the waiter that data is available.
+
+`waiter_register_notification` and `waiter_notify` can be called from different threads in any order.
+
+## Exposed API
+```c
+#define WAITER_RESULT_VALUES \
+    WAITER_RESULT_OK, \
+    WAITER_RESULT_CANCELLED, \
```

Comment:
>CANCELLED [](http://example.com/codeflow?start=18&length=9)

I see both CANCELLED and ABANDONED. Do we need both? - it's somewhat confusing.

Reply:
`WAITER_RESULT_CANCELLED` is for when the user explicitly cancels the operation. 

`WAITER_RESULT_ABANDONED` is for when the module is wants to close.

Reply:
I think they are slightly different cases, and the user might want to do differentiate between them.